becquerel/core/spectrum.py

```diff
@@ -983,6 +983,48 @@ def _mul_div(self, scaling_factor, div=False):
             spect_obj = Spectrum(bin_edges_raw=self.bin_edges_raw, **data_arg)
         return spect_obj
 
+    def attenuate(self, material, areal_density_gcm2: float, **kwargs):
+        """Compute a new Spectrum as if it were attenuated by some material.
+
+        Currently uses a lin-lin interpolation of the attenuation coefficient
+        (with coherent scattering) because XCOM's queries are limited to 100
+        energies.
+
+        Parameters
+        ----------
+        material : str | Iterable[str]
+            See tools.xcom.fetch_xcom_data
+        areal_density_gcm2 : float
+            Areal density of attenuating material in g/cm2. Negative values are
+            permitted, e.g., for estimating the unattenuated flux from an
+            observed attenuated spectrum.
+        **kwargs : dict
+            Additional parameters passed to self.__class__
+
+        Returns
+        -------
+        Spectrum
+        """
+        assert self.is_calibrated
+        e_min = max(1.0, self.bin_centers_kev[0])
+        e_max = min(1e8, self.bin_centers_kev[-1])
+        xd = tools.fetch_xcom_data(material, e_range_kev=(e_min, e_max))
+        mass_att_coeff = np.interp(
+            x=self.bin_centers_kev,
+            xp=xd["energy"],
+            fp=xd["total_w_coh"],
+        )
+        att = np.exp(-mass_att_coeff * areal_density_gcm2)
+        # the following is not (yet?) possible due to Spectrum.__mul__ limitations:
+        # return self * att
+
+        # instead, copy the pattern from Spectrum._mul_div:
+        if self._counts is not None:
+            data_arg = {"counts": self.counts * att}
```

Comment:
This statement leads to a `Spectrum` with noninteger counts, which we really shouldn't allow.

Reply:
In general we do allow noninteger counts data:
```py
class Spectrum:
    def __init__(self, ...):
    """
    ...
    Args:
          counts: counts per bin. array-like of ints, floats or UFloats, if
            uncs is not provided all values must be positive
    ...
    """
```

Reply:
Fair point. My worry here was that uncertainties would not be correctly estimated. I thought that no matter the type the counts are treated as if they were integers [when assigning uncertainties](https://github.com/lbl-anp/becquerel/blob/main/becquerel/core/spectrum.py#L113C1-L116C50):

```
          uncs (optional): iterable of uncertainty on the counts for each bin.
            If counts is given, is NOT a UFloat type, and uncs is not given,
            the uncertainties are assumed to be sqrt(N), with a minimum
            uncertainty of 1 (e.g. for 0 counts).
```

**However**, looking further into it, it looks like `self.counts` will in fact be a `uarray`, in which case `self.counts * att` is also a `uarray` with correctly-propagated uncertainties, and the [`handle_uncs` utility](https://github.com/lbl-anp/becquerel/blob/main/becquerel/core/utils.py#L48) would successfully detect this case and use the correct uncertainties, as indicated by this docstring.